component/http/v2/component.go

```diff
@@ -0,0 +1,99 @@
+// Package v2 provides a ready to use HTTP component.
+package v2
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/beatlabs/patron/log"
+)
+
+const (
+	port                = 50000
+	readTimeout         = 30 * time.Second
+	writeTimeout        = 60 * time.Second
+	idleTimeout         = 240 * time.Second
+	handlerTimeout      = 59 * time.Second // should be smaller than write timeout
+	shutdownGracePeriod = 5 * time.Second
+)
+
+// Component implementation of an HTTP router.
+type Component struct {
+	port                int
+	readTimeout         time.Duration
+	writeTimeout        time.Duration
+	shutdownGracePeriod time.Duration
+	handlerTimeout      time.Duration
+	handler             http.Handler
+	mu                  sync.Mutex
+	certFile            string
+	keyFile             string
+}
+
+func New(handler http.Handler, oo ...OptionFunc) (*Component, error) {
+	if handler == nil {
+		return nil, errors.New("handler is nil")
+	}
+
+	cmp := &Component{
+		port:                port,
+		readTimeout:         readTimeout,
+		writeTimeout:        writeTimeout,
+		shutdownGracePeriod: shutdownGracePeriod,
+		handlerTimeout:      handlerTimeout,
+		handler:             handler,
+	}
+
+	for _, option := range oo {
+		err := option(cmp)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return cmp, nil
+}
+
+// Run starts the HTTP server.
+func (c *Component) Run(ctx context.Context) error {
+	c.mu.Lock()
+	chFail := make(chan error)
+	srv := c.createHTTPServer()
+	go c.listenAndServe(srv, chFail)
+	c.mu.Unlock()
+
+	select {
+	case <-ctx.Done():
+		log.Info("shutting down HTTP component")
+		ctx, cancel := context.WithTimeout(context.Background(), c.shutdownGracePeriod)
+		defer cancel()
+		return srv.Shutdown(ctx)
```

Comment:
Will this return `ctx.Error()`? I am thinking that we should not have a case where returning because of `<-ctx.Done()` returns with `nil`.

Reply:
I do not follow, care to explain what you mean?

Reply:
Sure; this is a switch case for context cancelled. Is it guaranteed that it will `return ` with a `context.Canceled` error? Or will `srv.Shutdown(ctx)` return `nil` instead?

I am thinking that caller expects a non-nil error when things have been canceled (might be wrong though).

Reply:
@gmaz42  Actually the caller may receive both. In case of an nil, which happens in a shutdown, it will cancel the initial context which will notify all components to shutdown.
WDYT?

Reply:
What I am thinking is that if there was a cancellation -> it should be guaranteed to receive `ctx.Err()` back and not nil.

Do you agree with this or it is a stretch of how Patron components work in general? If it is a stretch then no problem, but in either cases perhaps we should spend a few GoDoc words explaining what happens in case of cancellation? (e.g. what is guaranteed and what not)

Reply:
The service.go orchestrates the all running components and monitors to receive wither an error or a nil.
An error is returned if the component cannot run for whatever reason e.g. kafka component tries to reconnect to the broker for a configured amount of time and then returns an error that is logger which will then trigger the shutting down of all other running components. 
All other components will gracefully shutdown without an error and that is exactly what the context cancellation does.
Only the original failure is logged the rest are shutdown and that is the reason we return nil in these cases.
So this is expected behavior. Makes sense?